internal/handlers/aggregate_events.go

```diff
@@ -0,0 +1,39 @@
+package handlers
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// AggregateEvents is called in the baseapp after a transaction has completed
+// This is used to modify the events that are emitted for a transaction.
+// anteEvents will be populated on failure and success
+// resultEvents will only be populated on success
+func AggregateEvents(anteEvents []abci.Event, resultEvents []abci.Event) ([]abci.Event, []abci.Event) {
+	if len(resultEvents) == 0 { // tx failed...fix fee event to have the exact fee charged
+		var txFee []byte
+		var feeIndex int
+		var feeFound, spenderFound bool
+		for i, event := range anteEvents {
+			if !feeFound && event.Type == sdk.EventTypeTx && string(event.Attributes[0].Key) == sdk.AttributeKeyFee {
```

Comment:
I feel like hard-coding the indexes in here is a recipe for future problems (talking about all the spots in this file where that's done).

Reply:
I agree.